solver/src/main/java/org/chocosolver/solver/constraints/nary/PropDiffN.java

```diff
@@ -36,95 +35,117 @@
+        for (int i = 0; i < n; i++) {
+            for (int j = i + 1; j < n; j++) {
+                if (mayOverlap(i, j)) {
+                    overlappingBoxes.addEdge(i, j);
+                }
+            }
+        }
+        pruneList = new TIntArrayList(n);
+    }
 
     //***********************************************************************************
     // METHODS
     //***********************************************************************************
 
     @Override
     public int getPropagationConditions(int idx) {
-        if (fast) return IntEventType.instantiation();
         return IntEventType.boundAndInst();
     }
 
-	@Override
-	public void propagate(int varIdx, int mask) throws ContradictionException {
-		int v = varIdx % n;
+    @Override
+    public void propagate(int varIdx, int mask) throws ContradictionException {
+        prop(varIdx);
+        forcePropagate(PropagatorEventType.CUSTOM_PROPAGATION);
+    }
+
+    private void prop(int varIdx) {
+        int v = varIdx % n;
         ISetIterator iter = overlappingBoxes.getNeighOf(v).iterator();
         while (iter.hasNext()) {
             int i = iter.nextInt();
-			if (!mayOverlap(v, i)) {
-				overlappingBoxes.removeEdge(v, i);
-			}
-		}
-		if (!boxesToCompute.contains(v)) {
-			boxesToCompute.add(v);
-		}
-		forcePropagate(PropagatorEventType.CUSTOM_PROPAGATION);
-	}
+            if (!mayOverlap(v, i)) {
+                overlappingBoxes.removeEdge(v, i);
+            }
+        }
+        if (!boxesToCompute.contains(v)) {
+            boxesToCompute.add(v);
+        }
+    }
 
     @Override
     public void propagate(int evtmask) throws ContradictionException {
-        if (PropagatorEventType.isFullPropagation(evtmask)) {
-            for (int i = 0; i < n; i++) {
-                overlappingBoxes.getNeighOf(i).clear();
-            }
-            for (int i = 0; i < n; i++) {
-                for (int j = i + 1; j < n; j++) {
-                    if (mayOverlap(i, j)) {
-                        overlappingBoxes.addEdge(i, j);
-                        if (boxInstantiated(i) && boxInstantiated(j)) {
-                            fails(); // TODO: could be more precise, for explanation purpose
+        boolean hasFiltered = true;
+        while(hasFiltered) {
+            hasFiltered = false;
+            if(PropagatorEventType.isFullPropagation(evtmask)) {
+                boxesToCompute.clear();
+                for (int i = 0; i < n; i++) {
+                    boxesToCompute.add(i);
+                    for (int j = i + 1; j < n; j++) {
+                        if (mayOverlap(i, j)) {
+                            overlappingBoxes.addEdge(i, j);
+                            if (boxInstantiated(i) && boxInstantiated(j)) {
+                                fails(); // TODO: could be more precise, for explanation purpose
+                            }
+                        } else {
+                            overlappingBoxes.removeEdge(i, j);
```

Comment:
Isn't that too strong ? There are cases where `mayOverlap` returns `false` and only one of the two conditions is falsified ?

Reply:
An edge in `overlappingBoxes` indicates that rectangle i and rectangle j might overlap in future instantiations and filterings, which should not happen.